### DISCUSSION TITLE
Fast github action compile with cache

### DISCUSSION THREAD

**Discussion by thewh1teagle** [cross-rs/cross#1352] [Q&A] [answered]
Is there a way to create github action to cross compile to multiple platforms with `cross` and have a cache for `target` so it will not recompile all the libraries every time?

**Comment by NobodyXu** [comment, marked as answer]
https://github.com/Swatinem/rust-cache can be used to cache target/ and it can be used with cross.
https://github.com/taiki-e/install-action can be used to install cross in GitHub Action

**Reply by thewh1teagle** [reply]
Thank you very much I tried so many things so many hours and this finally was good solution!

Do you know how can I use it to build for macos too?

Amyway if someone looking for working example I made you can see [here](https://github.com/thewh1teagle/oref-rs/blob/main/.github/workflows/build_server.yaml)

**Reply by NobodyXu** [reply]
AFAIK `cross` can't be used for MacOS and Windows cross building, though `rust-cache` and `install-action` would just work fine on these targets.

**Reply by thewh1teagle** [reply]
Small correction it does work for windows and linux but only for MacOS it doesn't

**Reply by NobodyXu** [reply]
I suppose it's because docker doesn't work on MacOS runner?
Not sure how to fix it.

**Reply by Emilgardis** [reply]
yeah macos github runners don't have docker unfortunately

One point with your example @thewh1teagle, [here](https://github.com/thewh1teagle/oref-rs/blob/aac5f6643aef8a8ff42e26d1386a22a204c0495c/.github/workflows/build_server.yaml#L380) you could use linux, the point of cross is to cross-compile to a target from any system that has rust, docker and can run linux/amd64 platforms containers

**Reply by thewh1teagle** [reply]
When I wrote this line I wondered few times which `os` to choose but then decided "not to take a chance" after all the tries I made
I will try it

**Reply by Emilgardis** [reply]
one thing with that though, it generally would be faster to just use the real system for compiling as cross imposes a container environment which can inccur some costs, especially if you're running on a system that is not x86_64

**Reply by NobodyXu** [reply]
Well if you only want to build/check without running tests/binary, then you can use cargo-zigbuild directly by `pip3 install cargo-zigbuild` then do `cargo zigbuild` or `cargo-zigbuild ...`, it won't need to pull any image.
